[YAMLParser] Unfold multi-line scalar values

igorkudrin opened the pull request:
Long scalar values can be split into multiple lines to improve readability. The rules are described in Section 6.5. "Line Folding", https://yaml.org/spec/1.2.2/#65-line-folding. In addition, for flow scalar styles, the Spec states that "All leading and trailing white space characters on each line are excluded from the content", https://yaml.org/spec/1.2.2/#73-flow-scalar-styles.

The patch implements these unfolding rules for double-quoted, single-quoted, and plain scalars.

Depends on #71774 and #71775

slinder1 reviewed (changes requested):
I don't mean to make existing debt your problem, but if it isn't too much work could you post a pre-patch that just adds the `FileCheck`s to the existing tests where the behavior changes, so the test diff is more self-documenting?

I left some comments but not so much about spec compliance as I don't have the time right now to work through it right now.

Either way, thank you for the work, and I'm happy to see the parser move towards more compliant parsing!

igorkudrin committed:
[YAMLParser] Unfold multi-line scalar values

Long scalar values can be split into multiple lines to improve
readability. The rules are described in Section 6.5. "Line Folding",
https://yaml.org/spec/1.2.2/#65-line-folding. In addition, for flow
scalar styles, the Spec states that "All leading and trailing white
space characters on each line are excluded from the content",
https://yaml.org/spec/1.2.2/#73-flow-scalar-styles.

The patch implements these unfolding rules for double-quoted,
single-quoted, and plain scalars.

igorkudrin commented on llvm/lib/Support/YAMLParser.cpp:
```diff
@@ -2030,187 +2030,219 @@ bool Node::failed() const {
+    if (UnquotedValue.size() == 1) {
+      Token T;
+      T.Range = UnquotedValue;
+      this->setError("Unrecognized escape code", T);
```
Fixed. Thanks.

igorkudrin commented on llvm/lib/Support/YAMLParser.cpp:
```diff
@@ -2030,187 +2030,219 @@ bool Node::failed() const {
-        break;
+  char LastNewLineAddedAs = '\0';
+  for (; I != StringRef::npos; I = UnquotedValue.find_first_of(LookupChars)) {
+    if (UnquotedValue[I] != '\x0D' && UnquotedValue[I] != '\x0A') {
```
It was an idea to be a bit closer to the spec, where all special characters are defined by their value. I changed them back to mnemonics in the last update.

igorkudrin commented on llvm/lib/Support/YAMLParser.cpp:
```diff
@@ -2030,187 +2030,219 @@ bool Node::failed() const {
+
+static StringRef
+parseScalarValue(StringRef UnquotedValue, SmallVectorImpl<char> &Storage,
+                 StringRef LookupChars,
```
Added a description for the function and its arguments.

igorkudrin commented:
> I don't mean to make existing debt your problem, but if it isn't too much work could you post a pre-patch that just adds the `FileCheck`s to the existing tests where the behavior changes, so the test diff is more self-documenting?

* Added #71774 for the tests
* Also added a unittest `YAMLParser.UnfoldsScalarValue` to check various combinations of line breaks and other characters. It seems like a `gtest`-based test suits better than a lit one.

slinder1 reviewed (approved):
Thank you for breaking out the NFC test changes, and for adding even more unittests!

I am again not being meticulous and exhaustive in looking for spec compliance, but in every case I did examine you were fixing a clear issue. I also ran a few cases through http://ben-kiki.org/ypaste/ and it agreed.

Thank you for the cleanup, and again thank you for taking all of my review comments on even when it required cleaning up an existing mess in the tests!
